### PR TITLE
RFC: Add support for dynamic doc strings

### PR DESCRIPTION
**Pull request by malmaud**
Useful for eg https://github.com/JuliaPy/PyCall.jl/issues/46.

Allows instances of the same type to have different docstrings which can only be determined at runtime. Useful for interlanguage support where the same Julia type (eg `PyObject`) corresponds to different types in the guest language that might have different docstrings.

### CODE REVIEW

File: base/docs/Docs.jl [hunk 1]
```diff
@@ -239,11 +239,26 @@ end
...
+
+If `getdoc` is defined on the value of the binding, then a dynamic docstring is returned instead of one based on the binding itself. 
 """
 function doc(binding::Binding, sig::Type = Union{})
+    if isdefined(binding.mod, binding.var)
+        var = getfield(binding.mod, binding.var)
+        if method_exists(getdoc, (typeof(var),))
+            return Base.Markdown.parse(getdoc(var))
+        end
+    end
```
Comment (MichaelHatherly) [inline base/docs/Docs.jl]: Can be a bit shorter/clearer as

```diff
 function doc(binding::Binding, sig::Type = Union{})
+    if defined(binding)
+        result = getdoc(resolve(binding), sig)
+        result === nothing || return Markdown.parse(result)
+    end
     results, groups = DocStr[], MultiDoc[]
```

with a default `getdoc(obj, sig) = nothing` method. Passing `sig` through to `getdoc` as well may be useful?

File: base/docs/Docs.jl [hunk 2]
```diff
@@ -285,7 +300,10 @@ end
 
 # Some additional convenience `doc` methods that take objects rather than `Binding`s.
 doc(obj::UnionAll) = doc(Base.unwrap_unionall(obj))
-doc(object, sig::Type = Union{}) = doc(aliasof(object, typeof(object)), sig)
+
+function doc(object, sig::Type = Union{})
+    doc(aliasof(object, typeof(object)), sig)
+end
```
Comment (MichaelHatherly) [inline base/docs/Docs.jl]: Unintentional formatting change?

**Review by MichaelHatherly — CHANGES_REQUESTED** [review]
Thanks @malmaud.

File: base/docs/Docs.jl (hunk 1 above)
Comment (malmaud) [inline base/docs/Docs.jl]: Good call.

File: base/docs/Docs.jl [hunk 3]
```diff
@@ -239,11 +239,25 @@ end
...
+"""
+getdoc(x, sig) = getdoc(x)
+getdoc(x) = nothing
+
+"""
     Docs.doc(binding, sig)
 
 Returns all documentation that matches both `binding` and `sig`.
+
+If `getdoc` returns a non-`nothing` result on the value of the binding, then a dynamic docstring is returned instead of one based on the binding itself.
```
Comment (tkelman) [inline base/docs/Docs.jl]: wrap long lines

Comment (malmaud) [inline base/docs/Docs.jl]: Done

**Comment by malmaud** [comment]
@MichaelHatherly look good now?

**Review by MichaelHatherly — APPROVED** [review]

File: base/docs/Docs.jl [hunk 4]
```diff
@@ -239,11 +239,27 @@ end
...
 
 Returns all documentation that matches both `binding` and `sig`.
+
+If `getdoc` returns a non-`nothing` result on the value of the binding, then a
+dynamic docstring is returned instead of one based on the binding itself.
 """
 function doc(binding::Binding, sig::Type = Union{})
+    if defined(binding)
+        result = getdoc(resolve(binding), sig)
+        result === nothing || return Markdown.parse(result)
```
Comment (MichaelHatherly) [inline base/docs/Docs.jl]: Would it make sense to *not* call `Markdown.parse` here and instead just return `result`? That way `getdocs` could return a more suitable object, such as an `RST` object for rst-formatted python docstrings.

Comment (malmaud) [inline base/docs/Docs.jl]: Ya, that's a good call.

File: base/docs/Docs.jl [hunk 5]
```diff
@@ -239,11 +239,27 @@ end
 # =================
 
 """
+    getdoc(x::T, sig) -> String
+
+Return the dynamic docstring associated with object `x`, or `nothing` to use
+the binding's documentation.
```
Comment (stevengj) [inline base/docs/Docs.jl]: The docstring doesn't explain what the `sig` argument is for.

**Pull request merged**